File: src/web/Controllers/StockListsController.cs
using System;
using System.Threading.Tasks;
using core.fs.Stocks.Lists;
using core.Shared;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using web.Utils;

namespace web.Controllers;

[ApiController]
[Authorize]
[Route("api/stocks/lists")]
public class StockListsController(Handler handler) : ControllerBase
{
    [HttpGet]
    public Task<ActionResult> StockLists() =>
        this.OkOrError(handler.Handle(new GetLists(User.Identifier())));

    [HttpPost]
    public Task<ActionResult> CreateStockList([FromBody]Create command) =>
        this.OkOrError(handler.HandleCreate(User.Identifier(), command));

    [HttpDelete("{id}")]
    public Task<ActionResult> DeleteStockList([FromRoute]Guid id) =>
        this.OkOrError(handler.Handle(new Delete(id, User.Identifier())));

    [HttpPost("{id}")]
    public Task UpdateStockList([FromBody]Update command) =>
        this.OkOrError(handler.HandleUpdate(command, User.Identifier()));

    [HttpPut("{id}")]
    public Task<ActionResult> AddStockToList([FromBody] AddStockToList command,
        [FromServices] Handler service) =>
        this.OkOrError(handler.HandleAddStockToList(User.Identifier(), command));

    [HttpDelete("{id}/{ticker}")]
    public Task<ActionResult> RemoveStockFromList([FromRoute] Guid id, [FromRoute] string ticker,
        [FromServices] Handler service) =>
        this.OkOrError(handler.Handle(new RemoveStockFromList(id, User.Identifier(), new Ticker(ticker))));

    [HttpPut("{id}/tags")]
    public Task<ActionResult> AddTagToStockList([FromBody]AddTagToList command) =>
        this.OkOrError(handler.HandleAddTagToList(User.Identifier(), command));

    [HttpDelete("{id}/tags/{tag}")]
    public Task<ActionResult> RemoveTagFromStockList([FromRoute] Guid id, [FromRoute] string tag) =>
        this.OkOrError(handler.Handle(new RemoveTagFromList(tag, id, User.Identifier())));

    [HttpGet("{id}")]
    public Task<ActionResult> GetStockList([FromRoute] Guid id) =>
        this.OkOrError(handler.Handle(new GetList(id, User.Identifier())));

    [HttpGet("{id}/export")]
    public Task<ActionResult> ExportStockList(Guid id, [FromQuery] bool justTickers) =>
        this.GenerateExport(
            handler.Handle(
                new ExportList(
                    justTickers: justTickers,
                    id: id,
                    userId: User.Identifier()
                )
            )
        );

    [HttpPost("{id}/clear")]
    public Task<ActionResult> ClearStockList([FromRoute] Guid id) =>
        this.OkOrError(handler.Handle(new Clear(id, User.Identifier())));
}

File: src/web.fs/Controllers/StockListsController.fs
namespace web.Controllers

open System
open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Authorization
open core.fs.Stocks.Lists
open core.Shared
open web.Utils

[<ApiController>]
[<Authorize>]
[<Route("api/stocks/lists")>]
type StockListsController(handler: Handler) =
    inherit ControllerBase()

    [<HttpGet>]
    member this.StockLists() =
        this.OkOrError(handler.Handle({GetLists.UserId = this.User.Identifier()}))

    [<HttpPost>]
    member this.CreateStockList([<FromBody>] command: Create) =
        this.OkOrError(handler.HandleCreate (this.User.Identifier()) command)

    [<HttpDelete("{id}")>]
    member this.DeleteStockList([<FromRoute>] id: Guid) =
        this.OkOrError(handler.Handle({Delete.Id = id; UserId = this.User.Identifier()}))

    [<HttpPost("{id}")>]
    member this.UpdateStockList([<FromBody>] command: Update) =
        this.OkOrError(handler.HandleUpdate command (this.User.Identifier()))

    [<HttpPut("{id}")>]
    member this.AddStockToList([<FromBody>] command: AddStockToList, [<FromServices>] service: Handler) =
        this.OkOrError(handler.HandleAddStockToList (this.User.Identifier()) command)

    [<HttpDelete("{id}/{ticker}")>]
    member this.RemoveStockFromList([<FromRoute>] id: Guid, [<FromRoute>] ticker: string, [<FromServices>] service: Handler) =
        this.OkOrError(handler.Handle({RemoveStockFromList.Id = id; UserId = this.User.Identifier(); Ticker = Ticker(ticker)}))

    [<HttpPut("{id}/tags")>]
    member this.AddTagToStockList([<FromBody>] command: AddTagToList) =
        this.OkOrError(handler.HandleAddTagToList (this.User.Identifier()) command)

    [<HttpDelete("{id}/tags/{tag}")>]
    member this.RemoveTagFromStockList([<FromRoute>] id: Guid, [<FromRoute>] tag: string) =
        this.OkOrError(handler.Handle({RemoveTagFromList.Tag = tag; Id = id; UserId = this.User.Identifier()}))

    [<HttpGet("{id}")>]
    member this.GetStockList([<FromRoute>] id: Guid) =
        this.OkOrError(handler.Handle({GetList.Id = id; UserId = this.User.Identifier()}))

    [<HttpGet("{id}/export")>]
    member this.ExportStockList(id: Guid, [<FromQuery>] justTickers: bool) =
        this.GenerateExport(
            handler.Handle(
                {ExportList.JustTickers = justTickers; Id = id; UserId = this.User.Identifier()}
            )
        )

    [<HttpPost("{id}/clear")>]
    member this.ClearStockList([<FromRoute>] id: Guid) =
        this.OkOrError(handler.Handle({Clear.Id = id; UserId = this.User.Identifier()}))
